# Q2 - Supervised Learning

In [1]:
# importing the libraries
# pandas for data manipulation and analysis
# numpy is using to manipulate multi-dimensional arrays and matrices. Also support mathematical functions

import numpy as np
import pandas as pd

## Loading the Data and Descriptive Analysis

In [2]:
# loading the data.
dataframe=pd.read_csv('quora_supervised.csv') # rading the data into a data frame.
len(dataframe)                                # returns number of rows in the dataframe.

200001

First we will read  all the data in the quora_supervised.csv file in to a pandas dataframe using.

The data set is consists of 200001.

In [3]:
# printing the first row
print(dataframe['question'][0])

How do I hack a wifi?


In [4]:
dataframe.head() # returns first 5 lines of dataframe

question  Topic number  \
0                            How do I hack a wifi?             0   
1                               Wwe is real fight?             6   
2                        What is written in vedas?             6   
3         Why don't women propose to men in India?             5   
4  Which car should I buy for my parents in India?             8   

       Topic desc  
0      Technology  
1  Work & Careers  
2  Work & Careers  
3        Politics  
4          Travel

In [5]:
dataframe.tail()  # returns last 5 lines of the dataframe

question  Topic number  \
199996  What is an inclined plane? What are examples o...             0   
199997  How can I prove that a negative number multipl...             5   
199998               How do I learn English step by step?             2   
199999  In what ways are you privileged and in what wa...             2   
200000                Is time travel possible? If yes how             6   

            Topic desc  
199996      Technology  
199997        Politics  
199998        Business  
199999        Business  
200000  Work & Careers

In [6]:
# Check for missing values
dataframe.isnull().sum()

question        0
Topic number    0
Topic desc      0
dtype: int64

There is no missing values in  the data set

In [7]:
# return count of each topic with in the data frame.
dataframe["Topic desc"].value_counts()

Learning          28192
Life Style        23038
Technology        21992
Work & Careers    19700
Politics          19659
Education         19143
Activities        18352
Relationships     18308
Travel            16653
Business          14964
Name: Topic desc, dtype: int64

All the topics in the data frame and the number of occurence of each topic is listed above

Learing is the most common and most frequent topic in the tadaset with the count of 28192 in total ie 14 % of the total data set. Life Style and Technology comes second and in the topic list. Business is the elast discussed topic in the platform.

## Split the data into train & test sets

In [42]:
# X contains data from question column
# y contains label data ie topic list
X = dataframe["question"]
y = dataframe["Topic desc"]

In [43]:
# Contains index and question text
X.head()

0                              How do I hack a wifi?
1                                 Wwe is real fight?
2                          What is written in vedas?
3           Why don't women propose to men in India?
4    Which car should I buy for my parents in India?
Name: question, dtype: object

In [44]:
# Contains index and topic name
y.head()

0        Technology
1    Work & Careers
2    Work & Careers
3          Politics
4            Travel
Name: Topic desc, dtype: object

In [45]:
# importing train_test_split from sklearn
from sklearn.model_selection import train_test_split

# test size represents the proportion of training and testing data split. 
# Random_state sets "randomness" of data randomisation
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=1)

Data is now split into training and testing data in 70:30 propotion

## Frequency transformation with Tf-idf

Using TfidfVectorizer from sklearn we can perform count vectorization and frequency transformation in one step. 

In [46]:
# importing TfidfVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

In [47]:
# initialise an instance of the TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer()

In [48]:
# Complete the vectorizing and fit transform on the original X_train
X_train_tfidf = tfidf_vectorizer.fit_transform(X_train)
# Examine shape of the dataset
X_train_tfidf.shape

(140000, 41761)

The doc_term_matrix is a sparse matrix of each word in all of the text data. Here the doc_term_matrix contains 200000 rows of questions data and 7155 columns or features corrsponding to rows. Adjusting the max_df and min_df values can increase or decrease the number of terms in the doc_term_matrix.

## 1. LinearSVC Model

LinearSVC is another implementation of Support Vector Classification for the case of a linear kernel. Note that LinearSVC does not accept keyword kernel, as this is assumed to be linear.Similar to SVC with parameter kernel=’linear’, but implemented in terms of liblinear rather than libsvm, so it has more flexibility in the choice of penalties and loss functions and should scale better to large numbers of samples.
This class supports both dense and sparse input and the multiclass support is handled according to a one-vs-the-rest scheme.

In [49]:
# importing LinearSVC model from sklearn.
from sklearn.svm import LinearSVC

In [50]:
# Contents of X_test
X_test.head()

140720    What is the worst thing that could happen in a...
194428    What type of government does Turkey have? How ...
5506                                     Can cats get AIDS?
38370     What is the best way to become a great program...
36930     What are the must see pieces of art at the Uff...
Name: question, dtype: object

In [51]:
# Create an instance of the LinearSVC classifier
classifier = LinearSVC()

# model training
classifier.fit(X_train_tfidf, y_train)

LinearSVC(C=1.0, class_weight=None, dual=True, fit_intercept=True,
     intercept_scaling=1, loss='squared_hinge', max_iter=1000,
     multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
     verbose=0)

In [53]:
# Transform original test question data to a vector
X_test_transform = tfidf_vectorizer.transform(X_test)

# Predict message type from Linear SVC classifier
predictions = classifier.predict(X_test_transform)

In [54]:
predictions.shape

(60001,)

The predicted array is a list of 60001 topics.

## confusion matrix

In [56]:
# Show a confusion matrix of results
from sklearn import metrics
print(metrics.confusion_matrix(y_test, predictions))

[[4522   59  134  122  127  108  141  120   82  120]
 [  72 3823   78  103   54   71   91   95   55   87]
 [ 108   45 4873   97  112  114  106  101   99  133]
 [  87   55  105 7633   78   58   99  110   85  127]
 [ 120   46   81  113 6041   91   97  124   88  106]
 [ 103   41   90   80  107 5068   96   99   85   68]
 [ 100   64  101  172  127   96 4566   75   97   89]
 [  97   55   98  136  141  100   89 5719   80   88]
 [ 101   70  105  107   96   73   89   82 4233   77]
 [  90   75  150  119  105   82  130   91   88 4915]]


In [57]:
## Print a classification report based on confusion matrix
print(metrics.classification_report(y_test,predictions))

                precision    recall  f1-score   support

    Activities       0.84      0.82      0.83      5535
      Business       0.88      0.84      0.86      4529
     Education       0.84      0.84      0.84      5788
      Learning       0.88      0.90      0.89      8437
    Life Style       0.86      0.87      0.87      6907
      Politics       0.86      0.87      0.87      5837
 Relationships       0.83      0.83      0.83      5487
    Technology       0.86      0.87      0.87      6603
        Travel       0.85      0.84      0.84      5033
Work & Careers       0.85      0.84      0.84      5845

     micro avg       0.86      0.86      0.86     60001
     macro avg       0.86      0.85      0.85     60001
  weighted avg       0.86      0.86      0.86     60001



Precision is the  the percentage of the correct predictions.It predicts if data is of a particular class
Here Precision for this model is ranges from .83 to .88. Business and Learning are the two predicted models with highest Precision. The micro and macro avg of the Precision is .86. The topic label with lowest precision is Relationships with value of .83.

Recall is the number of correct results divided by the number of results that should have been returned. It is the ability of a model to find all relevant instances in a dataset. Here Learning has the highest recall rate - .90, followed by Life Style, Politics and Technology with recall value of .87. The label Activities only have recall rate of 0.82.

F1-score is the harmonic mean of the precision and recall. It provides optimum balance between precision and recal. It shows how precise and robust a classifier is. It value ranges between (0,1). Here the weighted avg of f1-score is 0.86. The topic Learning has the highest fi-score and Relationships and Activities has the minimum value. The over all f1- score of this modes is ranges from 0.83 to 0.89.


Support is the number of occurrences of each class.

In [24]:
print(metrics.accuracy_score(y_test,predictions))

0.8565357244045932


The accuracy defines the percentage of times the model correctly predicted the class. Only works well on dataset with  equal number of samples in each class.It is defined as number of correct predictions made by a model divided bytotal predictions.

For this model the accuracy is 86.65 % which is a pretty good accuracy for this dataset.

## Predicting 10 Random Questions

In [62]:
def predict_topic_type(text_message):
    transformed_message = tfidf_vectorizer.transform([text_message])
    model_output = classifier.predict(transformed_message)
    return(model_output)

In [71]:
# Submit a text message to the function and see if the model can accurately detect whether it is ham or spam.
# Here I'm testing whether the model can predict this as SPAM

predict_topic_type("What do my recent dreams of love mean?")

array(['Life Style'], dtype=object)

In [78]:
import random
random_qes = random.sample(range(1, 200000), 10)
print(random_qes)

[174559, 175088, 70919, 48976, 39618, 94880, 6609, 183751, 59478, 116493]


In [79]:
count = 0 
for i in random_qes:
    print('\n')
    print(f"Question-{count} : {dataframe['question'][i]}")
    print(f"Right answer : {dataframe['Topic desc'][i]}")
    print(f"Predicted Topic :{predict_topic_type(dataframe['question'][i])}")
    count = count+1
    print ('-------------------------------------------')



Question-0 : What is a remainder?
Right answer : Education
Predicted Topic :['Education']
-------------------------------------------


Question-1 : How cold can the Gobi Desert get, and how do its average temperatures compare to the ones in the Sahara?
Right answer : Business
Predicted Topic :['Business']
-------------------------------------------


Question-2 : Is Dissociative Identity Disorder hereditary?
Right answer : Technology
Predicted Topic :['Technology']
-------------------------------------------


Question-3 : What kind of economy does the United States have compared to the economies of other countries?
Right answer : Relationships
Predicted Topic :['Relationships']
-------------------------------------------


Question-4 : How can I Remotely hack a mobile phone?
Right answer : Technology
Predicted Topic :['Technology']
-------------------------------------------


Question-5 : Will the value of rupee increase after the 500 and 1000 rupees notes are banned?
Right answer

In [ ]:
Here topic of 10 out of 10 questions are predicted correctly.

# Some Additional Models

As part of this CA, i build some other models also. Programing code  with confusion matrix is adding below. 
I selected LinearSVC Model model as best fit model because its high accuracy, fi-value, Precision and Recall.



### MultinomialNB (model -2)

The multinomial Naive Bayes classifier is suitable for classification with discrete features (e.g., word counts for text classification). The multinomial distribution normally requires integer feature counts. However, in practice, fractional counts such as tf-idf may also work.

In [81]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.feature_extraction.text import CountVectorizer

In [82]:
count_vectorizer = CountVectorizer(max_df=0.90, min_df=4, stop_words="english")
doc_term_matrix = count_vectorizer.fit_transform(dataframe["question"])
target_topic = dataframe['Topic number']

In [83]:
X_train, X_test, y_train, y_test = train_test_split(doc_term_matrix,target_topic, test_size = 0.3, random_state = 1)

In [84]:
mnc = MultinomialNB()
mnc.fit(X_train, y_train)

MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True)

In [85]:
model_predictions = mnc.predict(X_test)

In [86]:
print(metrics.confusion_matrix(y_test,model_predictions))

[[5171  268   40  146  137  192   80   96   73  400]
 [ 150 5938   42  134  103  113   88   88   76  175]
 [ 104  126 3317   90  138  118  155  110   64  307]
 [ 153  251   49 3999  140  160  166  277   89  251]
 [  82  207   71  115 4123  146  149  137   81  376]
 [ 119  216   50  115  141 4734  128   99   81  154]
 [ 147  206   62  120  135  103 4537  201   85  249]
 [ 101  215   44  105  112  262  174 4469  127  179]
 [  91  200   84   99  134  144  108  159 3818  196]
 [ 209  109   58   61  110   65  140  178   97 7410]]


In [87]:
print(metrics.classification_report(y_test, model_predictions))

              precision    recall  f1-score   support

           0       0.82      0.78      0.80      6603
           1       0.77      0.86      0.81      6907
           2       0.87      0.73      0.79      4529
           3       0.80      0.72      0.76      5535
           4       0.78      0.75      0.77      5487
           5       0.78      0.81      0.80      5837
           6       0.79      0.78      0.78      5845
           7       0.77      0.77      0.77      5788
           8       0.83      0.76      0.79      5033
           9       0.76      0.88      0.82      8437

   micro avg       0.79      0.79      0.79     60001
   macro avg       0.80      0.78      0.79     60001
weighted avg       0.79      0.79      0.79     60001



In [88]:
print(metrics.accuracy_score(y_test,model_predictions))

0.7919201346644222


The Overall accuracy of the model is 79.19 

### Logistic Regression (Model - 3)

Logistic regression, despite its name, is a linear model for classification rather than regression. Logistic regression is also known in the literature as logit regression, maximum-entropy classification (MaxEnt) or the log-linear classifier. In this model, the probabilities describing the possible outcomes of a single trial are modeled using a logistic function. Logistic regression is implemented in LogisticRegression. This implementation can fit binary, One-vs-Rest, or multinomial logistic regression with optional  or Elastic-Net regularization.

In [94]:
from sklearn.linear_model import LogisticRegression

In [95]:
logistic_reg_model = LogisticRegression(solver='lbfgs')
logistic_reg_model.fit(X_train, y_train)
logistic_reg_model_predictions = logistic_reg_model.predict(X_test)

/home/shyam/anaconda3/envs/AI2_course/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)
/home/shyam/anaconda3/envs/AI2_course/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)
/home/shyam/anaconda3/envs/AI2_course/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)
/home/shyam/anaconda3/envs/AI2_course/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:758: ConvergenceWarning: lbfgs failed to converge. Increase the number of iterations.
  "of iterations.", ConvergenceWarning)
/home/shyam/anaconda3/envs/AI2_course/lib/python3.7/site-pack

In [96]:
print(metrics.confusion_matrix(y_test, logistic_reg_model_predictions))
print(metrics.classification_report(y_test,logistic_reg_model_predictions))

[[5727  141   45  109   88  104   97   75   67  150]
 [ 144 5934   47  127  124  103  114   89   88  137]
 [ 129   70 3713   76  105   75   91   68   58  144]
 [ 136  146   68 4420  145  122  128  133   97  140]
 [  91  132   73  115 4457  113  104  111  104  187]
 [ 122  123   45  109  118 4930   85  107  111   87]
 [ 127  106   76  102  153  101 4795  143   95  147]
 [ 109  126   56  117  130  119  163 4746  101  121]
 [ 101  111   63   90  116   93   91  132 4107  129]
 [ 144   81   63   93  126   63  120  105   82 7560]]
              precision    recall  f1-score   support

           0       0.84      0.87      0.85      6603
           1       0.85      0.86      0.86      6907
           2       0.87      0.82      0.85      4529
           3       0.82      0.80      0.81      5535
           4       0.80      0.81      0.81      5487
           5       0.85      0.84      0.85      5837
           6       0.83      0.82      0.82      5845
           7       0.83      0.82   

In [97]:
print(metrics.accuracy_score(y_test,logistic_reg_model_predictions))

0.8398026699555008


The overall accuracy of the model is 83.98%
Which is less than the LinearSVC Model but larger than the MultinomialNB.

# References:

Metrics and Scoring: Quantifying the Quality of Predictions — Scikit-Learn 0.22.2 Documentation [online] (2020) available: https://scikit-learn.org/stable/modules/model_evaluation.html [accessed 12 Apr 2020].

Sklearn.Model_selection.Train_test_split — Scikit-Learn 0.22.2 Documentation [online] (2020) available: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html [accessed 12 Apr 2020].

Sklearn.Decomposition.LatentDirichletAllocation — Scikit-Learn 0.22.2 Documentation [online] (2020) available: https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.LatentDirichletAllocation.html [accessed 12 Apr 2020].

sklearn.svm.LinearSVC — Scikit-Learn 0.22.2 Documentation [online] (2020) available: https://scikit-learn.org/stable/modules/generated/sklearn.svm.LinearSVC.html [accessed 12 Apr 2020].


Sklearn.Feature_extraction.Text.CountVectorizer — Scikit-Learn 0.22.2 Documentation [online] (2020) available: https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html [accessed 14 Apr 2020].

Sklearn.Model_selection.GridSearchCV — Scikit-Learn 0.22.2 Documentation [online] (2020) available: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html [accessed 14 Apr 2020].


Sklearn.Naive_bayes.MultinomialNB — Scikit-Learn 0.22.2 Documentation [online] (2020) available: https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html [accessed 15 Apr 2020].


sklearn.linear_model.LogisticRegression — Scikit-Learn 0.22.2 Documentation [online] (2020) available: https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html [accessed 15 Apr 2020].

API Reference — Pandas 1.0.3 Documentation [online] (2020) available: https://pandas.pydata.org/docs/reference/index.html [accessed 15 Apr 2020].
